### ISSUE TITLE
Steps towards dissolving spack fork

### ISSUE THREAD

**Issue by AlexanderRichert-NOAA** [JCSDA/spack-stack#436]
Following our 10 Jan 2023 discussion, this issue is for updating spack-stack/spack fork so as to reduce differences between our spack fork and main spack, with the ultimate goal of eliminating the need for our own fork. The following is intended as a comprehensive running checklist of items to resolve:

- [x] Move jcsda-emc spack extensions to spack-stack (done in #961)
- [x] Move jcsda-emc spack repos to spack-stack  (done in #961)
- [x] Add extra instructions functionality for container recipes in main spack (waiting on main spack devs; see https://github.com/spack/spack/pull/32478)
- ~~Determine whether module path/hash length-related modifications are still needed~~ maintain this difference for now
- ~~Resolve package differences:~~ captured in another issue

**Comment by climbfuji** [comment]
Do we actually believe we'll be able to completely dissolve our spack fork?

**Comment by AlexanderRichert-NOAA** [comment]
Yes. Not necessarily in the 1.4.0 release, but, yes.

**Comment by climbfuji** [comment]
@AlexanderRichert-NOAA Do you still believe that this is possible? If so, for which release?

**Comment by AlexanderRichert-NOAA** [comment]
I thought we had reworked this issue, but maybe I'm misremembering. We can see where things stand after the next merge. Given that we have to periodically make and test updates to whichever Spack repo we end up using, the use of the fork doesn't necessarily create a lot more workload. What would you think of moving JCSDA/EMC-specific stuff (the lib/ subdir and the repo dirs) to the spack-stack repository, so we would be using a clean or nearly clean copy of Spack? I don't think it's critical, but the main benefit would be that we could swap in clones of main Spack for testing purposes without having to do a bunch of shuffling things around. I think the idea would be to have steps in setup.sh like what I experimented with a while back to point Spack to the appropriate extensions & spack repos.

**Comment by climbfuji** [comment]
I remember we had to make some changes to `spack/lib` as well for our extension to the spack command and for other things. If we could split `spack` into a more or less clean hash from the authoritative repo with the same name and a separate directory `spack-ext`, then that would be great.

**Comment by AlexanderRichert-NOAA** [comment]
Yeah that sounds good. When we/I do the next merge from main Spack I can work on this as well and hopefully get that working for the next release.

**Comment by climbfuji** [comment]
@AlexanderRichert-NOAA Any suggestion on how to proceed with this? To me it feels like this should be turned into an epic or project and broken up into individual manageable pieces that we can tackle one at a time as time permits and as we think makes sense?

**Comment by AlexanderRichert-NOAA** [comment]
If you want we can take the checklist and break it out into separate issues? I think pulling the jcsda-emc lib/ and repos/ subdirectories should be reasonably straightforward. I'm not as certain about the whole container build issue.

As far as our wanting to not have hashes in the modulefile directory structure, is that a strict requirement, or did we just remove them because they're ugly/unnecessary? When I talked to the Spack devs about it some months ago, it sounded like there was some openness to letting users set the hash length, but, could we potentially live without that modification?

**Comment by climbfuji** [comment]
> If you want we can take the checklist and break it out into separate issues? I think pulling the jcsda-emc lib/ and repos/ subdirectories should be reasonably straightforward. I'm not as certain about the whole container build issue.

Maybe there's a way around the container build issue if we can specify our own template (and the location of it) and leave the original template in place. We can consider this in a later step?

> As far as our wanting to not have hashes in the modulefile directory structure, is that a strict requirement, or did we just remove them because they're ugly/unnecessary? When I talked to the Spack devs about it some months ago, it sounded like there was some openness to letting users set the hash length, but, could we potentially live without that modification?

I would be very surprised if that was acceptable by the users - it is certainly ugly. We can and should ask our users. I would say let's try to convince the spack-developers to make this configurable, our changes are trivial and not intrusive and do exactly that.

**Comment by climbfuji** [comment]
Resolved (in parts, rest will be addressed elsewhere) by https://github.com/JCSDA/spack-stack/pull/961

**Comment by climbfuji** [comment]
We may have to revert this, since it breaks the spack container builds.
